### PR TITLE
Dev data import for needs and shelters from live API

### LINKED ISSUE #20: TODO: Add import from live API to help developers
We used to be able to run `rails google:import` and get a full local database, but we'll get out of that business soon as #16 and #17 land

So, let's create a rake task:

* Pulls all needs from api.harveyneeds.org/api/v1/needs and then creates local needs
* Pulls all shelters from  api.harveyneeds.org/api/v1/shelters and then creates local shelters

Something like this with the `http` gem

```
Http.get("https://api.harveyneeds.org/api/v1/needs").body.to_json["needs"].each do |need|
  Need.create! need
end
```

### PR DESCRIPTION
**Pull request by nihonjinrxs**
Closes #20.

I used HTTParty instead of HTTP, since it has a nicer interface to create a class around the response.

### CODE REVIEW

File: app/helpers/api_importer.rb
Comment (ironcoconut) [inline app/helpers/api_importer.rb]: These methods should live in the job. Added them as a helper means they get included in the views which we don't need. It is a "feature" of rails. 

You could put it in models, but these are just helper methods and we don't need them outside of the job.

File: app/jobs/import_needs_job.rb
Comment (ironcoconut) [inline app/jobs/import_needs_job.rb:11]: Does this account for duplicate records? Or is it meant to only be run when starting a new device env?

If only Dev, can we get comments here and in the readme to that effect? I don't want this ever accidently run in production.

**Review by ironcoconut — CHANGES_REQUESTED** [review]
Is the auto imported disabled? We don't want to run this in production.

Tests?

File: app/jobs/import_needs_job.rb
Comment (nihonjinrxs) [reply app/jobs/import_needs_job.rb:11]: I could make it account for dups, but the issue ( #20 ) indicated for developer help.

Let me take a look and see what I can do.

Comment (nihonjinrxs) [reply app/jobs/import_needs_job.rb:11]: Looks like the current imports just delete all data first, then import everything, which is already pretty heavy-handed.

See:

* https://github.com/sketch-city/harvey-api/blob/master/app/jobs/import_needs_job.rb#L19
* https://github.com/sketch-city/harvey-api/blob/master/app/jobs/import_shelters_job.rb#L19

File: app/helpers/api_importer.rb
Comment (jwo) [reply app/helpers/api_importer.rb]: I'm probably cool with this being it's own class, but not in `helpers`... how about an `app/importers` class?

File: app/jobs/import_needs_job.rb
Comment (ironcoconut) [reply app/jobs/import_needs_job.rb:11]: Dev only then. Makes good sense.

File: app/helpers/api_importer.rb
Comment (ironcoconut) [reply app/helpers/api_importer.rb]: +1

File: app/jobs/import_needs_job.rb
Comment (jwo) [reply app/jobs/import_needs_job.rb:11]: @ironcoconut this is for dev only, setup only, one time only. Basically only in a `db:seed`

File: app/helpers/api_importer.rb
Comment (nihonjinrxs) [reply app/helpers/api_importer.rb]: I was torn on that.  I'd rather have both methods in a common class (why implement the same class twice?), so I can't put it in the same file as the jobs, because then it would unavailable to the other job.  Should I just move this file out of `app/helpers` and into `app/jobs`?

File: app/jobs/import_needs_job.rb
Comment (ironcoconut) [reply app/jobs/import_needs_job.rb:11]: My bad. I forgot that we discussed this earlier.

**Comment by nihonjinrxs** [comment]
Regarding the auto-run question, @jwo's PR #17 removes all the Google Sheets import-related stuff, including the app hooks and endpoints. We discussed in Slack merging that one first.  This one creates a replacement Rake task, but no hooks or endpoints for production use.  Happy to add some docs if that helps.

On tests, I wasn't sure how best to test this one.  Happy to write something, but may need some help in that direction.

File: app/helpers/api_importer.rb
Comment (nihonjinrxs) [reply app/helpers/api_importer.rb]: I can move this to an `app/importers` folder.

File: app/jobs/import_needs_job.rb
Comment (nihonjinrxs) [reply app/jobs/import_needs_job.rb:11]: @ironcoconut @jwo Did you want me to add a `delete_all` before importing in each job, like the Google Sheets one has?

**Comment by ironcoconut** [comment]
Got it on PR #17. Will removing that code create a merge conflict here?

For tests, I think just run it and make sure it does not error and that it adds records. You can wrap it in a VCR block so that it runs faster.

Do we have a no merge label? I'm on mobile.

Looks good.

**Comment by nihonjinrxs** [comment]
OK, can add a test that does that.  Never used VCR -- what would that look like?

**Comment by ironcoconut** [comment]
I think the gem is in the repo.

https://github.com/vcr/vcr/blob/master/README.md

**Comment by jwo** [comment]
For VCR, I think tests might be overboard honestly.  But if your curious, I use VCR tests to prevent the Test suite from actually calling out to
Amazon each test run. 

@ironcoconut -- will you take this pull and merge when you're happy?

I do think we should add to the README on how devs should use this. 

I would also be cool removing all classes and just moving code into db/seeds. I don't want people to think it's more important than it is.

**Comment by nihonjinrxs** [comment]
Added basic tests that run the import and check that the number of records added to the database match what you get back from the API.  Used VCR to capture and store the request data.

Happy to move this into seeds if that makes more sense, but I think that would also need some explanation, since you don't generally need an internet connection when setting up the database for an app.  Seems better to have an explicit task to run for that.

Anyway, let me know what else you'd like for me to do.

**Comment by nihonjinrxs** [comment]
Ah, crap. Travis couldn't load `HTTParty`? Not getting that error locally when I run tests. Any ideas?

**Comment by nihonjinrxs** [comment]
OK, not sure what's happening here. Help appreciated. @jwo @ironcoconut

**Comment by jwo** [comment]
@nihonjinrxs that's _weird_ ... I'll take a look and see if I can figure out

File: app/importers/api_importer.rb
Comment (jwo) [inline app/importers/api_importer.rb]: @nihonjinrxs 
I'm pretty sure this should be `require 'httparty'`

it's working on your mac (I assume) because mac's hard drive's are case insentive, but in prod on linux its case sensitive.

File: app/jobs/import_needs_job.rb
Comment (ironcoconut) [reply app/jobs/import_needs_job.rb:11]: I think it's ok as is.

**Comment by ironcoconut** [comment]
Looks good. If you could also add a short blurb to the readme about how to run and why, it would be appreciated. I'll approve once the tests pass.

**Comment by ironcoconut** [comment]
@jwo do we need to wait for your 2 prs before merging?

**Comment by jwo** [comment]
@ironcoconut No, we can merge this in and then update the other PRs to keep them up to date.

File: app/importers/api_importer.rb
Comment (nihonjinrxs) [reply app/importers/api_importer.rb]: Ah, gotcha.  Will mod and see if that fixes it.

**Comment by nihonjinrxs** [comment]
Also added some info to the README.  Build is passing.  Let me know if more is needed.  Thanks!

cc @ironcoconut @jwo

**Review by ironcoconut — DISMISSED** [review]

**Comment by jwo** [comment]
@nihonjinrxs Will you make one final change here? There's a conflict in Gemfile that'll be easy to solve.

Make your Gemfile look like this.. you're adding redis. (that's already in master, but git can't figure out what to do)

```
gem 'connection_pool'
gem 'redis', '~> 3.2'
gem "httparty", "~> 0.15.6"

group :development, :test do
```

@nihonjinrxs UPDATE: I could use the github editor and solve. \o/ .. no need for you to do anything.

**Review by jwo — APPROVED** [review]
Thanks for making this happen @nihonjinrxs , Appreciate your contributions!

**Comment by nihonjinrxs** [comment]
Sorry I missed the last mention there, glad you could fix it. Happy to help!  Looking at #21 next.

**Pull request merged**
